## Rubric

Instructions: DELETE this cell before you submit via a `git push` to your repo before deadline. This cell is for your reference only and is not needed in your report. 

Scoring: Out of 10 points

- Each Developing  => -2 pts
- Each Unsatisfactory/Missing => -4 pts
  - until the score is 

If students address the detailed feedback in a future checkpoint they will earn these points back


|                  | Unsatisfactory                                                                                                                                                                                                    | Developing                                                                                                                                                                                              | Proficient                                     | Excellent                                                                                                                              |
|------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------|
| Data relevance   | Did not have data relevant to their question. Or the datasets don't work together because there is no way to line them up against each other. If there are multiple datasets, most of them have this trouble | Data was only tangentially relevant to the question or a bad proxy for the question. If there are multiple datasets, some of them may be irrelevant or can't be easily combined.                       | All data sources are relevant to the question. | Multiple data sources for each aspect of the project. It's clear how the data supports the needs of the project.                         |
| Data description | Dataset or its cleaning procedures are not described. If there are multiple datasets, most have this trouble                                                                                              | Data was not fully described. If there are multiple datasets, some of them are not fully described                                                                                                      | Data was fully described                       | The details of the data descriptions and perhaps some very basic EDA also make it clear how the data supports the needs of the project. |
| Data wrangling   | Did not obtain data. They did not clean/tidy the data they obtained.  If there are multiple datasets, most have this trouble                                                                                 | Data was partially cleaned or tidied. Perhaps you struggled to verify that the data was clean because they did not present it well. If there are multiple datasets, some have this trouble | The data is cleaned and tidied.                | The data is spotless and they used tools to visualize the data cleanliness and you were convinced at first glance                      |


# COGS 108 - Data Checkpoint

## Authors

Instructions: REPLACE the contents of this cell with your team list and their contributions. Note that this will change over the course of the checkpoints

This is a modified [CRediT taxonomy of contributions](https://credit.niso.org). For each group member please list how they contributed to this project using these terms:
> Analysis, Background research, Conceptualization, Data curation, Experimental investigation, Methodology, Project administration, Software, Visualization, Writing – original draft, Writing – review & editing

Example team list and credits:
- Alice Anderson: Conceptualization, Data curation, Methodology, Writing - original draft
- Bob Barker:  Analysis, Software, Visualization
- Charlie Chang: Project administration, Software, Writing - review & editing
- Dani Delgado: Analysis, Background research, Visualization, Writing - original draft

## Research Question

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback



## Background and Prior Work

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Hypothesis


Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback


## Data

### Data overview

Instructions: REPLACE the contents of this cell with descriptions of your actual datasets.

For each dataset include the following information
- Dataset #1
  - Dataset Name:
  - Link to the dataset:
  - Number of observations:
  - Number of variables:
  - Description of the variables most relevant to this project
  - Descriptions of any shortcomings this dataset has with repsect to the project
- Dataset #2 (if you have more than one!)
  - same as above
- etc

Each dataset deserves either a set of bullet points as above or a few sentences if you prefer that method.

If you plan to use multiple datasets, add a few sentences about how you plan to combine these datasets.

In [ ]:
# Run this code every time when you're actively developing modules in .py files.  It's not needed if you aren't making modules
#
## this code is necessary for making sure that any modules we load are updated here 
## when their source code .py files are modified

%load_ext autoreload
%autoreload 2

In [21]:
# Setup code -- this only needs to be run once after cloning the repo!
# this code downloads the data from its source to the `data/00-raw/` directory
# if the data hasn't updated you don't need to do this again!

# if you don't already have these packages (you should!) uncomment this line
# %pip install requests tqdm

import sys
sys.path.append('./modules') # this tells python where to look for modules to import

import get_data # this is where we get the function we need to download data
import pandas as pd
import numpy as np

# replace the urls and filenames in this list with your actual datafiles
# yes you can use Google drive share links or whatever
# format is a list of dictionaries; 
# each dict has keys of 
#   'url' where the resource is located
#   'filename' for the local filename where it will be stored 
#datafiles = [
#    { 'url': 'https://raw.githubusercontent.com/fivethirtyeight/data/refs/heads/master/airline-safety/airline-safety.csv', 'filename':'airline-safety.csv'},
#    { 'url': 'https://raw.githubusercontent.com/fivethirtyeight/data/refs/heads/master/bad-drivers/bad-drivers.csv', 'filename':'bad-drivers.csv'}
#]

#get_data.get_raw(datafiles,destination_directory='data/00-raw/')

### Dataset #1 

Instructions: 
1. Change the header from Dataset #1 to something more descriptive of the dataset
2. Write a few paragraphs about this dataset. Make sure to cover
   1. Describe the important metrics, what units they are in, and giv some sense of what they mean.  For example "Fasting blood glucose in units of mg glucose per deciliter of blood.  Normal values for healthy individuals range from 70 to 100 mg/dL.  Values 100-125 are prediabetic and values >125mg/dL indicate diabetes. Values <70 indicate hypoglycemia. Fasting idicates the patient hasn't eaten in the last 8 hours.  If blood glucose is >250 or <50 at any time (regardless of the time of last meal) the patient's life may be in immediate danger"
   2. If there are any major concerns with the dataset, describe them. For example "Dataset is composed of people who are serious enough about eating healthy that they voluntarily downloaded an app dedicated to tracking their eating patterns. This sample is likely biased because of that self-selection. These people own smartphones and may be healthier and may have more disposable income than the average person.  Those who voluntarily log conscientiously and for long amounts of time are also likely even more interested in health than those who download the app and only log a bit before getting tired of it"
3. Use the cell below to 
    1. load the dataset 
    2. make the dataset tidy or demonstrate that it was already tidy
    3. demonstrate the size of the dataset
    4. find out how much data is missing, where its missing, and if its missing at random or seems to have any systematic relationships in its missingness
    5. find and flag any outliers or suspicious entries
    6. clean the data or demonstrate that it was already clean.  You may choose how to deal with missingness (dropna of fillna... how='any' or 'all') and you should justify your choice in some way
    7. You will load raw data from `data/00-raw/`, you will (optionally) write intermediate stages of your work to `data/01-interim` and you will write the final fully wrangled version of your data to `data/02-processed`
4. Optionally you can also show some summary statistics for variables that you think are important to the project
5. Feel free to add more cells here if that's helpful for you


### The Condemned: Death Sentences in the Modern Era (1976–2019)

This dataset, compiled by *The Intercept*, includes all known individuals sentenced to death in active U.S. death penalty jurisdictions from July 2, 1976 (the date of *Gregg v. Georgia*) through November 30, 2019. It contains **7,335 individuals**, with each row representing one person sentenced to death during the modern death penalty era.

As of 2019:
- 1,448 individuals have been executed  
- 2,752 remain on death row  
- 3,135 have been removed from death row for reasons other than execution (commutations, exonerations, death in custody, etc.)

This dataset allows us to examine how demographic, geographic, and temporal factors are associated with different death row outcomes.

---

### Important Limitations

1. The dataset only includes individuals who received a death sentence. It does not include individuals charged with capital crimes who did not receive the death penalty, so we cannot analyze disparities in who receives a death sentence.

2. The data were compiled from corrections departments, public defenders, advocacy groups, and independent trackers. Reporting standards may vary across states, which may introduce inconsistencies.

3. Individuals sentenced earlier have had more time for appeals, commutations, or execution. Recent sentences may appear more often as “currently on death row” due to time effects.

4. The outcome categories are legally complex and require careful grouping for meaningful statistical analysis.

---


In [25]:
# Load the dataset
file_path = "data/00-raw/the-condemed-data.csv"
df = pd.read_csv(file_path)

In [26]:
# Dataset details
print("Shape of dataset:", df.shape)
print("\nColumn names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

Shape of dataset: (7335, 21)

Column names:
['Status', 'state', 'sentencing_date', 'sentencing_year', 'full_name', 'last_name', 'first_name', 'race', 'County', 'DOB', 'gender', 'currently_on_death_row', 'executed', 'volunteer_execution', 'died', 'suicide', 'commuted', 'exonerated', 'resentenced', 'current_sentence', 'released']

Data types:
Status                     object
state                      object
sentencing_date            object
sentencing_year           float64
full_name                  object
last_name                  object
first_name                 object
race                       object
County                     object
DOB                        object
gender                     object
currently_on_death_row     object
executed                   object
volunteer_execution        object
died                       object
suicide                    object
commuted                   object
exonerated                 object
resentenced                object
current_sen

### Key Variables

| Variable | Type | Units / Format | Why It Matters |
|-----------|------|----------------|----------------|
| `first_name` | Text | String | Name component (can support matching/validation, but not unique). |
| `last_name` | Text | String | Name component (can support matching/validation, but not unique). |
| `status` | Categorical | Text label (e.g., "Currently on Death Row", "Executed") | High-level case status that summarizes the individual’s current legal situation and can be cross-validated with binary outcome indicators. |
| `race` | Categorical | Race classification | Used to examine racial disparities in death row outcomes. |
| `gender` | Categorical | Male / Female | Demographic characteristic. |
| `dob` | Date | YYYY-MM-DD (after parsing) | Used to calculate age at sentencing and analyze demographic patterns. |
| `sentencing_date` | Date | Mixed formats (full date, month/year, or year-only) | Used to analyze temporal trends; may require cleaning due to inconsistent formatting. |
| `sentencing_year` | Numeric | Year (e.g., 1977) | Clean and consistent time variable used for trend analysis. |
| `state` | Categorical | U.S. state (2-letter code) | Captures geographic variation in sentencing and outcomes. |
| `current_sentence` | Categorical | DR, LWOP, LPP, Y, M, TIME, etc. | Primary detailed outcome variable used to categorize death row outcomes. |


**Key for `current_sentence`:**
- **DR** – Currently on Death Row  
- **LWOP** – Life Without Parole  
- **LPP** – Life With Possibility of Parole  
- **Y** – Term of Years  
- **M** – Term of Months  
- **TIME** – Time Served  
- Additional categories indicate execution, exoneration, resentencing, or death in custody  

The `current_sentence` field serves as the main outcome variable and will be grouped into broader categories (executed, still on death row, sentence reduced, exonerated, died, etc.) for analysis.

---

In [27]:
# Preview first few rows
df.head()

,Status,state,sentencing_date,sentencing_year,full_name,last_name,first_name,race,County,DOB,...,currently_on_death_row,executed,volunteer_execution,died,suicide,commuted,exonerated,resentenced,current_sentence,released
0,Not Currently on Death Row,AL,08/17/1976,1976.0,Donald Thigpen,Thigpen,Donald,Black,Escambia,1952,...,N,N,NaN,NaN,NaN,NaN,NaN,Y,LPP,NaN
1,Not Currently on Death Row,AL,11/22/1976,1976.0,Jerry Wayne Jacobs,Jacobs,Jerry,White,Blount,1949,...,N,N,NaN,NaN,NaN,NaN,NaN,Y,LWOP,NaN
2,Not Currently on Death Row,AL,11/22/1976,1976.0,John Lewis Jacobs,Jacobs,John,White,Blount,NaN,...,N,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Not Currently on Death Row,AL,1977,1977.0,James Willie Cochran,Cochran,James,Black,Jefferson,1943,...,N,N,NaN,NaN,NaN,NaN,Y,Y,DR,Y
4,Not Currently on Death Row,AL,1977,1977.0,Tommy Lewis,Lewis,Tommy,Black,Coffee,1951,...,N,N,NaN,NaN,NaN,NaN,NaN,Y,LWOP,NaN


In [28]:
# Standardize column names
df.columns = df.columns.str.strip().str.lower().str.replace(" ", "_")

print("Updated column names:")
print(df.columns.tolist())


# Convert date columns to datetime
df['sentencing_date'] = pd.to_datetime(df['sentencing_date'], errors='coerce')
df['dob'] = pd.to_datetime(df['dob'], errors='coerce')


# Convert binary indicator columns to boolean
binary_cols = [
    'currently_on_death_row',
    'executed',
    'volunteer_execution',
    'died',
    'suicide',
    'commuted',
    'exonerated',
    'resentenced',
    'released'
]

for col in binary_cols:
    df[col] = df[col].astype(str).str.strip().str.lower().map({
        'true': True,
        'false': False,
        '1': True,
        '0': False,
        'yes': True,
        'no': False,
        'y': True,
        'n': False
    })

print("\nUpdated data types:")
print(df.dtypes)


# Check duplicates
print("\nDuplicate rows:", df.duplicated().sum())


# Verify one row per individual
print("\nUnique individuals:", df['full_name'].nunique())
print("Total rows:", len(df))


Updated column names:
['status', 'state', 'sentencing_date', 'sentencing_year', 'full_name', 'last_name', 'first_name', 'race', 'county', 'dob', 'gender', 'currently_on_death_row', 'executed', 'volunteer_execution', 'died', 'suicide', 'commuted', 'exonerated', 'resentenced', 'current_sentence', 'released']

Updated data types:
status                            object
state                             object
sentencing_date           datetime64[ns]
sentencing_year                  float64
full_name                         object
last_name                         object
first_name                        object
race                              object
county                            object
dob                       datetime64[ns]
gender                            object
currently_on_death_row            object
executed                          object
volunteer_execution               object
died                              object
suicide                           object
commuted       

First, column names were standardized by removing extra spaces, converting all names to lowercase, and replacing spaces with underscores. This ensures consistent referencing throughout the analysis.

Next, the `sentencing_date` and `dob` columns were converted to datetime format to allow proper time-based calculations and trend analysis.

Binary indicator columns (such as `executed`, `commuted`, `exonerated`, etc.) were converted to boolean values (True/False). This ensures consistent formatting and allows for accurate logical comparisons and summary statistics.

The dataset was then checked for duplicate rows to confirm that no records were repeated.

Finally, the number of unique individuals (based on `full_name`) was compared to the total number of rows to verify that each row represents a distinct case.


In [30]:
# Find duplicated names
duplicate_names = df[df.duplicated('full_name', keep=False)].sort_values('full_name')

duplicate_names[['full_name', 'state', 'sentencing_date', 'current_sentence']].head(10)


,full_name,state,sentencing_date,current_sentence
7327,Alfredo Prieto,VA,NaT,NaN
1279,Alfredo Prieto,CA,1992-06-16,NaN
4517,Alton Coleman,OH,1985-05-06,NaN
3187,Alton Coleman,IN,1986-04-12,NaN
6090,Charles Evans,TX,1979-04-18,LPP
3191,Charles Evans,IN,1986-09-19,180Y
1869,Darnell Williams,CA,2016-09-21,NaN
3194,Darnell Williams,IN,1987-03-23,LWOP
3556,David Brown,LA,2018-06-22,NaN
3551,David Brown,LA,2011-10-29,NaN


We checked for duplicated values in the `full_name` column to determine whether any individuals appear more than once in the dataset. While some names appear multiple times, these records differ by state and/or sentencing date.

This indicates that these are different individuals who happen to share the same name, rather than duplicate records of the same person. Therefore, `full_name` is not a unique identifier, but each row still represents a distinct case.


In [31]:
# Missing values count
missing_counts = df.isna().sum().sort_values(ascending=False)

# Missing percentage
missing_percent = (df.isna().mean() * 100).sort_values(ascending=False)

missing_summary = pd.DataFrame({
    "Missing Count": missing_counts,
    "Missing %": missing_percent
})

missing_summary


,Missing Count,Missing %
commuted,7267,99.072938
suicide,7232,98.595774
exonerated,7203,98.200409
volunteer_execution,7203,98.200409
released,7002,95.460123
died,6743,91.929107
dob,5911,80.586230
resentenced,4963,67.661895
current_sentence,4938,67.321063
sentencing_date,574,7.825494


### Missing Data Interpretation

Most missingness in this dataset occurs in binary event indicator columns such as `commuted`, `suicide`, `exonerated`, `volunteer_execution`, `released`, and `died`. These columns have very high percentages of missing values (over 90%). This is expected and not a data quality issue. These variables only contain values when the event occurred, so missing values generally mean the event did not happen.

The `dob` column has about 80% missing values. However, missingness does not appear to differ meaningfully between executed and non-executed individuals, suggesting it is not strongly related to outcome. This likely reflects inconsistent historical reporting rather than systematic bias. Because of this, age-based analysis may be limited.

Core analytic variables such as `state`, `race`, `status`, `sentencing_year`, `gender`, and `executed` have almost no missing values. These are the primary variables used in our analysis, and their completeness supports reliable outcome and demographic comparisons.

Overall, missingness is largely structural (due to event-based columns) and does not significantly compromise the main analyses of demographic and geographic patterns in death row outcomes.


In [33]:
# Sentencing date before 1976 (should not happen)
df[df['sentencing_date'].dt.year < 1976]


,status,state,sentencing_date,sentencing_year,full_name,last_name,first_name,race,county,dob,...,currently_on_death_row,executed,volunteer_execution,died,suicide,commuted,exonerated,resentenced,current_sentence,released


In [34]:
# Future sentencing dates (should not happen)
df[df['sentencing_date'] > pd.Timestamp.today()]


,status,state,sentencing_date,sentencing_year,full_name,last_name,first_name,race,county,dob,...,currently_on_death_row,executed,volunteer_execution,died,suicide,commuted,exonerated,resentenced,current_sentence,released


In [35]:
# Unrealistic situation 1
df[(df['executed'] == True) & (df['currently_on_death_row'] == True)]


,status,state,sentencing_date,sentencing_year,full_name,last_name,first_name,race,county,dob,...,currently_on_death_row,executed,volunteer_execution,died,suicide,commuted,exonerated,resentenced,current_sentence,released


In [36]:
# Unrealistic situation 2
df[(df['exonerated'] == True) & (df['executed'] == True)]


,status,state,sentencing_date,sentencing_year,full_name,last_name,first_name,race,county,dob,...,currently_on_death_row,executed,volunteer_execution,died,suicide,commuted,exonerated,resentenced,current_sentence,released


We checked for unrealistic or logically impossible entries in the dataset.

First, we verified that no sentencing dates occurred before 1976 (the beginning of the modern death penalty era) and that there were no future sentencing dates. Both checks returned zero rows, indicating no temporal inconsistencies.

Next, we checked for logical contradictions in outcome variables. Specifically, we confirmed that no individual was marked as both executed and currently on death row, and that no one was marked as both executed and exonerated. These checks also returned zero rows.

Overall, no obvious outliers or contradictory records were detected, suggesting that the dataset is internally consistent.


In [37]:
df['age_at_sentencing'] = (
    df['sentencing_date'].dt.year - df['dob'].dt.year
)

df['age_at_sentencing'].describe()


count    1256.000000
mean       31.112261
std         9.829185
min       -17.000000
25%        24.000000
50%        29.000000
75%        36.000000
max        77.000000
Name: age_at_sentencing, dtype: float64

In [38]:
# Unrealistic ages (under 15 or over 90)
suspicious_age = df[(df['age_at_sentencing'] < 15) | 
                    (df['age_at_sentencing'] > 90)]

suspicious_age[['full_name', 'state', 'age_at_sentencing']].head()


,full_name,state,age_at_sentencing
1440,Karl Darnell Holmes,CA,-17.0
4866,Benjamin Brewer,OK,10.0


In [39]:
# Remove clearly impossible ages only if they exist
df = df[~((df['age_at_sentencing'] < 15) | 
          (df['age_at_sentencing'] > 90))]

print("Shape after removing impossible ages:", df.shape)


Shape after removing impossible ages: (7333, 22)


We calculated age at sentencing by subtracting the birth year from the sentencing year. We then checked for unrealistic ages, specifically individuals under 15 years old or over 90 years old at the time of sentencing.

Two suspicious cases were identified, including one negative age and one age of 10, which are clearly impossible. These records were removed to maintain data accuracy.

After removing these unrealistic values, the dataset contained 7,333 observations.

Summary statistics show that among valid cases (1,256 individuals with available age data), the average age at sentencing was approximately 31 years old.

Overall, age values appear reasonable after cleaning, although many observations are missing age information.


In [40]:
df = df.dropna(how='all')

In [41]:
df.to_csv("data/02-processed/condemned_cleaned.csv", index=False)

In [42]:
file_path = "data/02-processed/condemned_cleaned.csv"
df_clean = pd.read_csv(file_path)

df_clean.head()

,status,state,sentencing_date,sentencing_year,full_name,last_name,first_name,race,county,dob,...,executed,volunteer_execution,died,suicide,commuted,exonerated,resentenced,current_sentence,released,age_at_sentencing
0,Not Currently on Death Row,AL,1976-08-17,1976.0,Donald Thigpen,Thigpen,Donald,Black,Escambia,1952-01-01,...,False,NaN,NaN,NaN,NaN,NaN,True,LPP,NaN,24.0
1,Not Currently on Death Row,AL,1976-11-22,1976.0,Jerry Wayne Jacobs,Jacobs,Jerry,White,Blount,1949-01-01,...,False,NaN,NaN,NaN,NaN,NaN,True,LWOP,NaN,27.0
2,Not Currently on Death Row,AL,1976-11-22,1976.0,John Lewis Jacobs,Jacobs,John,White,Blount,NaN,...,False,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Not Currently on Death Row,AL,NaN,1977.0,James Willie Cochran,Cochran,James,Black,Jefferson,1943-01-01,...,False,NaN,NaN,NaN,NaN,True,True,DR,True,NaN
4,Not Currently on Death Row,AL,NaN,1977.0,Tommy Lewis,Lewis,Tommy,Black,Coffee,1951-01-01,...,False,NaN,NaN,NaN,NaN,NaN,True,LWOP,NaN,NaN


After completing all cleaning and validation steps, we removed any rows that were completely empty (if any existed). 

The cleaned dataset was then saved to `data/02-processed/condemned_cleaned.csv` to preserve a finalized version for analysis. 

Finally, we reloaded the saved file to confirm that it was written correctly. The cleaned dataset contains 7,333 observations and 22 variables, indicating that only clearly impossible age entries were removed.


### Dataset #2 

See instructions above for Dataset #1.  Feel free to keep adding as many more datasets as you need.  Put each new dataset in its own section just like these. 

Lastly if you do have multiple datasets, add another section where you demonstrate how you will join, align, cross-reference or whatever to combine data from the different datasets

Please note that you can always keep adding more datasets in the future if these datasets you turn in for the checkpoint aren't sufficient.  The goal here is demonstrate that you can obtain and wrangle data.  You are not tied down to only use what you turn in right now.

In [ ]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE


## Ethics

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback

## Team Expectations 

Instructions: REPLACE the contents of this cell with your work, including any updates to recover points lost in your proposal feedback


## Project Timeline Proposal

Instructions: Replace this with your timeline.  **PLEASE UPDATE your Timeline!** No battle plan survives contact with the enemy, so make sure we understand how your plans have changed.  Also if you have lost points on the previous checkpoint fix them